bpi.module

```diff
@@ -857,3 +555,207 @@ function bpi_error_message($exception, $level) {
   // Send the exception to logger.
   watchdog_exception('bpi', $exception);
 }
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ *
+ * Add custom submit handler for showing BPI specific fields and saving BPI
+ * data in db and add validate handler, which will add our own submit handler.
+ */
+function bpi_form_ding_news_node_form_alter(&$form, &$form_state, $form_id) {
+  // Don't show BPI fields if node is new.
+  if (!empty($form['nid']) || !$form['nid']['#value']) {
+    unset($form['field_bpi_workflow']);
+    return;
+  }
+  // Don't show BPI fields if user doesn't have access to the workflow tab.
+  $node = node_load($form['nid']['#value']);
+  if (!workflow_tab_access("node", $node)) {
+    unset($form['field_bpi_workflow']);
+    return;
+  }
+
+  // Get the 'Sent to BPI' (Save and push) state id.
+  $workflows = workflow_get_workflow_names();
+  $workflows_reversed = array_flip($workflows);
+  $wid = $workflows_reversed['BPI'];
+  $states = workflow_get_workflow_state_names($wid);
+  $states_reversed = array_flip($states);
+  $push_sid = $states_reversed['Sent to BPI'];
+  $delete_sid = $states_reversed['Deleted from BPI'];
+  $created_sid = $states_reversed['Created from BPI'];
+
+  // Remove 'Save and push' button if the current state is already 'Sent to
+  // BPI'. We have to do this, because it's not possible to disallow workflow
+  // transitions to the current state.
+  if ($form['field_bpi_workflow'][LANGUAGE_NONE][0]['#default_value'] == $push_sid) {
+    unset($form['field_bpi_workflow'][LANGUAGE_NONE][0]['workflow']['workflow_sid']['#options'][$push_sid]);
+  }
+
+  // Remove "Delete from BPI" button, when we have already deleted from BPI.
+  if ($form['field_bpi_workflow'][LANGUAGE_NONE][0]['#default_value'] == $delete_sid) {
+    unset($form['field_bpi_workflow'][LANGUAGE_NONE][0]['workflow']['workflow_sid']['#options'][$delete_sid]);
+  }
+
+  $nid = $form['nid']['#value'];
+  // Only add BPI fields when the node exists, it hasn't just been created from
+  // BPI, and the workflow state isn't "Created from BPI".
+  if ($nid > 0 && !isset($_GET['bpi_id']) && $form['field_bpi_workflow'][LANGUAGE_NONE][0]['#default_value'] != $created_sid) {
+    $form['bpi'] = bpi_http_push_action_form($nid ? $nid : 0);
```

Comment:
How can the value to bpi_http_push_action_form be 0 when we just checked that $nid > 0 in the if statement above?